content/blog/2020-09-11-futil-vs-fsm/index.md

```diff
@@ -0,0 +1,285 @@
+middle of the high-level input programs that imply an abstract architecture,
+and the low-level Verilog programs that instantiate a concrete architecture.
+By concretizing some aspects of the final architecture, while keeping others
+abstract, FuTIL is able to implement optimizations that can be shared across
+many different toolchains.
+
+**FuTIL programs**. FuTIL programs are constructed using *components* which
+roughly correspond to functions in the software world, and modules in the
+hardware world.
+Each component has three sections:
+(1) *cells* which describes all hardware units used by the component,
+(2) *wires* which describes connections between the hardware units, and
+(3) *control* which describes the execution schedule of the program.
+We elide the specifics of the first two sections. For a detailed overview,
+please take a look at the [FuTIL documentation][futil].
+
+The *control* program is the novel aspect of FuTIL's design. A control program
+looks like:
+```C
+seq {
+  init;
+  while lt.out with cmp {
+    incr;
+  }
+}
+```
+Here, `seq` and `while` are two *control-flow operators*. They describe the
+execution flow of the program. `init`, `incr`, and `cmp` are dataflow graphs
+are defined in the *wires* section of the component. At a high-level, they
+define an "action". For example, `init` might define the connections required
+to initialize a register, `incr` for incrementing the value in the register,
+and `cmp` defines the connections to compare the current value in the register.
+Using these groups, the control program precisely describes the control flow.
+
+The key insight in the design of FuTIL is that by explicitly representing
+the control flow of a program, the mid-level IR can implement several
+optimizations that would otherwise not be possible in a Verilog designs.
+For example, if the control program says that two dataflow graphs execute
+in sequence, we know that they can share the same hardware resource and that
+they will never conflict.
+
+
+### The Dahlia-to-FuTIL Compiler
+
+[Dahlia][] is an imperative, array-based programming language that is used
+to generate hardware accelerators.
+Dahlia programs look very similar to simple C/C++ programs. For example,
+the following implements a dot-product kernel:
+
+```C
+decl a: float[8];
+decl b: float[8];
+let sum = 0;
+for (let i = 0 .. 8) {
+  sum += a[i] * b[i];
+}
+```
+
+The first two lines define two memories `a` and `b` and the `for` loop iterates
+over their elements and accumulates the values into a register `sum`.
+
+The Dahlia compiler compiles Dahlia programs into FuTIL in three steps:
+(1) type checking, which establishes that the generated design will have
+predictable area-latency trade-offs, (2) lowering which removes complex
+control structures like `for` loops, loop unrolling, and memory banking, and
+(3) code generation which instantiates hardware components and generates
+FuTIL programs.
```

Comment:
Maybe add commas after "lowering" and "code generation" to match the one after "type checking."